Advanced Filters: Add url support

psealock opened the pull request:
Allow Advanced Filters to initialise from url query parameters and manipulate parameters as a function of changes in the UI.

Given an added filter,

![screen shot 2018-09-04 at 2 43 46 pm](https://user-images.githubusercontent.com/1922453/45007231-f808d200-b050-11e8-883e-892252ee16a0.png)

A URL representation will be `&status_is=pending`

![screen shot 2018-09-04 at 2 44 48 pm](https://user-images.githubusercontent.com/1922453/45007325-63eb3a80-b051-11e8-85ad-d32b1af01ec4.png)

Filters with multiple values, such as products, will have ids for products joined by ",". Encoded, it will look like this in the url

```
&product_includes=50%2C48
```

 ### Tasks to be completed in follow up PRs

* Handling tags in Search component. I'll need to query the products endpoint in a follow up PR. Until then, i'm just using the id as the label.
![screen shot 2018-09-04 at 3 12 22 pm](https://user-images.githubusercontent.com/1922453/45008124-f9d49480-b054-11e8-9703-17dfe1196651.png)
* Add url support for `&match=any`
* Look into allowing the `,` to not be encoded/decoded so it remains in the url. Is it worth it? (I think yes)

### Test

1. `/wp-admin/admin.php?page=wc-admin#/analytics/orders?filter=advanced`
2. Add/remove/modify/clear filters and see the url update accordingly

psealock committed:
Search: move selected from state to props

justinshreve commented:
> Does the encoded value matter? Is it worth not decoding the "+" so it looks like this in the url
&product_excludes=50+48 instead of &product_excludes=50%2B48?

Should these be comma separated instead? It looks like GA and others do this when there is multiple selection in the URL. Example: `u.date00=20161002&_u.date01=20180903&_.useg=builtin1,builtin27,builtin102,builtin7`

I don't think it's a big deal "visually" if the items are encoded. My assumption is that users will be generating the URLs with the UI and sharing them that way. Anyone who is changing the URL manually will probably know what url encoding is.

> Does introducing query parameters such as below make sense?

Yes I think so. We'll want to make those filters sharable for sure.

I suggest the `product_includes` naming compared to `products_includes`.

ryelle commented:
> Does the encoded value matter? Is it worth not decoding the "+" so it looks like this in the url
&product_excludes=50+48 instead of &product_excludes=50%2B48?
> Does introducing query parameters such as below make sense?

What about [following core's lead on this?](https://codex.wordpress.org/Class_Reference/WP_Query#Post_.26_Page_Parameters) In WP_Query, you can say `post__in=2,3,4` to get posts 2, 3, and 4; or `post__not_in=2,3,4` to get [default query] without posts 2, 3, and 4.

Some WP_Query params also accept negative numbers as a "exclusion", so we could have `products_is`/`products_is_not` use the same query param (`product`) and flip the sign (whether this is a good thing to mimic I'm not sure, but it's a suggestion).

```
products_includes -> product__in=[comma-sep IDs]
products_excludes -> product__not_in=[comma-sep IDs]
products_is       -> product=[comma-sep IDs]
products_is_not   -> product=["negative" comma-sep IDs]
```

We might also need to follow up with Proton about the API, their order report endpoint only mentions a single param `products=24,120`, to "Limit result set to all items that have the specified product assigned." - no mention of the `Is` vs `Includes` functionality (see p7bje6-14E-p2) @timmyc 

I noticed the URL is updated immediately, rather than after clicking the Filter button, will that be fixed before this is ready for "real" review?

psealock commented:
Thanks for the helpful input @ryelle @justinshreve !

psealock commented on client/components/filters/index.js:
```diff
@@ -48,6 +49,7 @@ const ReportFilters = ( { advancedConfig, filters, query, path } ) => {
 		case 'advanced':
 			advancedCard = (
 				<AdvancedFilters
+					key={ queryKey }
```
I added this so that internal state is "reset" on change of the url. The downside is the `<Search />` internal state is reset as well, losing recent search results. If we'd like to fix this, I can use `getDerivedStateFromProps` instead or keep results of the search in the data layer for quick access.

psealock commented:
Going to refactor a bit... moving back to In Progress

psealock commented:
Ready for review, thanks

timmyc commented on client/components/filters/advanced/index.js:
```diff
@@ -127,9 +99,15 @@ class AdvancedFilters extends Component {
 		const filterConfig = this.props.config[ key ];
-		const newFilter = { key, rule: filterConfig.rules[ 0 ] };
+		const newFilter = { key };
+		if ( filterConfig.rules ) {
```
Do we need to ensure this is an array that has length here so we can access the object safely below?

timmyc commented on client/components/filters/advanced/search-filter.js:
```diff
@@ -0,0 +1,61 @@
+		const { filter, config, onFilterChange } = this.props;
+		const { key, rule, value } = filter;
+		const selected = value.map( id => {
+			// For now
```
What else needs to be done here / what is blocking it?

timmyc replied:
NM saw this:

> Handling tags in Search component. I'll need to query the products endpoint in a follow up PR. Until then, i'm just using the id as the label.

timmyc commented on client/components/filters/advanced/search-filter.js:
```diff
@@ -0,0 +1,61 @@
+	}
+}
+
+export default SearchFilter;
```
Missing proptypes/documentation.

timmyc commented on client/components/filters/advanced/select-filter.js:
```diff
@@ -0,0 +1,37 @@
+	);
+};
+
+export default SelectFilter;
```
proptypes again.

timmyc commented on client/components/filters/advanced/test/utils.js:
```diff
@@ -0,0 +1,179 @@
+		const query = getQueryFromActiveFilters( activeFilters );
+		expect( query.status_is ).toBe( 'open' );
+		expect( query.things_includes ).toBe( '1,2,3' );
+		expect( query.customer ).toBe( 'new' );
```
Very detailed tests, thank you.

timmyc commented on client/components/filters/advanced/utils.js:
```diff
@@ -0,0 +1,115 @@
+ * @return {string} - url query key.
+ */
+export const getUrlKey = ( key, rule ) => {
+	if ( rule ) {
```
Should we check rule.length?

timmyc commented on client/components/filters/advanced/utils.js:
```diff
@@ -0,0 +1,115 @@
+ * @return {array} - array of Search values
+ */
+export const getSearchFilterValue = str => {
+	return str.length ? str.split( ',' ) : [];
```
should we `trim()` before length on this one?

timmyc commented on client/components/filters/advanced/utils.js:
```diff
@@ -0,0 +1,115 @@
+	return str.length ? str.split( ',' ) : [];
+};
+
+/**
```
Appears to be an orphaned doc block?

psealock replied:
Naw, its a typedef, http://usejsdoc.org/tags-typedef.html, for `activeFilter` since its a common shape I use throughout the file

timmyc reviewed (commented):
This is looking quite nice Paul! Left a few comments, and I did notice that during testing, I was able to add a new filter like a coupon code even though there wasn't a match found, which results in the duplicate PUSH error:

![filters](https://user-images.githubusercontent.com/22080/45319825-c3d05e00-b4f5-11e8-8a9a-83359730300c.gif)

I'm fine with tackling some validation on items like that in a follow-up though.

psealock commented:
Thanks for the review @timmyc 

* Fixed various checks for arrays, array length, string length.
* Trimmed strings before placing in the url.
* Added proptypes where needed.
* Changed the `disabled` logic on the Filter button so you shouldn't be able to click it if the url won't update.

> I was able to add a new filter like a coupon code even though there wasn't a match found

I think this is on the radar for `<Search />` as it progresses

I also fiddled with some padding so the focus styles after adding a filter look a bit better:

<img width="750" alt="screen shot 2018-09-11 at 2 56 27 pm" src="https://user-images.githubusercontent.com/1922453/45336924-c7d5ac00-b5d9-11e8-8a1d-8ddb042fd5d1.png">

timmyc reviewed (approved):
Testing out well for me still. LGTM

timmyc commented:
Actually just found something, but you can handle in a follow - up. The Filter button is disabled on the coupon scenario, but it is still clickable, which results in the PUSH state error, and the clear all filters link is kinda jumpy when clicking the button too:

![button-clickable-still](https://user-images.githubusercontent.com/22080/45397420-c6fe4380-b5f4-11e8-9485-a4db4315845b.gif)

So fine to merge this still, or if you want to fix in this branch I'm happy to test again if you would like me to.
